### PR TITLE
[CIT-450] Add tests for plugin_sm lib

### PR DESCRIPTION
**Pull request by makr11st**
## Proposed changes

Add tests for sm_agent and its components

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: sm/plugins/tedge_dummy_plugin/tests/tests.rs [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
Comment (makr11st) [inline sm/plugins/tedge_dummy_plugin/tests/tests.rs:1]: This is a trick to get the `tedge_dummy_plugin` build with all tests such it can be used in tests.

**Review by abelikt — APPROVED** [review]
Looks good so far. Two little concerns about the --version and --module-version still open but these shouldn't block this here.

File: sm/plugin_sm/tests/plugin.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        let mut file = tempfile::Builder::new()
+            .suffix(".0")
+            .tempfile_in(path)
+            .unwrap();
+
+        let content = r#"{"name":"abc","version":"1.0"}"#;
+        let _a = file.write_all(content.as_bytes()).unwrap();
+
+        let res = plugin.list().await;
+        assert!(res.is_ok());
```
Comment (rina23q) [inline sm/plugin_sm/tests/plugin.rs:50]: I think it's better to check the content of Ok().

File: sm/plugin_sm/tests/plugin.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        let _a = file.write_all(content.as_bytes()).unwrap();
+
+        let module = SoftwareModule {
+            module_type: Some("test".into()),
+            name: "test".into(),
+            version: None,
+            url: None,
+        };
+        let res = plugin.install(&module).await;
+        assert!(res.is_ok());
```
Comment (rina23q) [inline sm/plugin_sm/tests/plugin.rs:73]: Very minor. It's technically the same as line 30. I like assert_eq as it's easier to understand res is Ok(()).
```
assert_eq!(res, Ok(()));
```

File: sm/plugin_sm/tests/plugin.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        let _a = file.write_all(content.as_bytes()).unwrap();
+
+        let module = SoftwareModule {
+            module_type: Some("test".into()),
+            name: "test".into(),
+            version: None,
+            url: None,
+        };
+        let res = plugin.remove(&module).await;
+        assert!(res.is_ok());
```
Comment (rina23q) [inline sm/plugin_sm/tests/plugin.rs:96]: Same as above.

File: sm/plugin_sm/tests/plugin_manager.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    #[test]
+    fn plugin_manager_load_plugins_some_by_plugins_none() {
+        let temp_dir = tempfile::tempdir().unwrap();
+        let _file = create_some_plugin_in(&temp_dir);
+        let plugin_dir = temp_dir.path().to_owned();
+        let mut plugins = ExternalPlugins::open(plugin_dir).unwrap();
+        let _ = plugins.load();
+
+        assert!(plugins.by_software_type("test").is_none());
+        assert!(plugins.by_file_extension("test").is_none());
```
Comment (rina23q) [inline sm/plugin_sm/tests/plugin_manager.rs:38]: Is it impossible that the random plugin file's type is "test" or extension is "test"?

File: sm/plugins/tedge_dummy_plugin/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "tedge_dummy_plugin"
+version = "0.0.0"
```
Comment (rina23q) [inline sm/plugins/tedge_dummy_plugin/Cargo.toml:3]: :laughing:

**Review by rina23q — APPROVED** [review]
Only minor comments.

File: sm/plugin_sm/tests/plugin_manager.rs (hunk 5 above)
Comment (makr11st) [reply sm/plugin_sm/tests/plugin_manager.rs:38]: That's correct.

File: sm/plugin_sm/tests/plugin.rs (hunk 2 above)
Comment (makr11st) [reply sm/plugin_sm/tests/plugin.rs:62]: Done.

File: sm/plugin_sm/tests/plugin.rs (hunk 3 above)
Comment (makr11st) [reply sm/plugin_sm/tests/plugin.rs:93]: If there is no errors the API return only `Ok(())`.

File: sm/plugin_sm/tests/plugin.rs (hunk 4 above)
Comment (makr11st) [reply sm/plugin_sm/tests/plugin.rs:123]: As above, no errors, no additional output.

**Pull request merged**